File: VB/WpfApplication25/ViewModel.vb
Imports System.Collections.ObjectModel
Imports System.ComponentModel

Namespace WpfApplication25

    Public Class ViewModel

        Public Property List As ObservableCollection(Of TestData)

        Public Sub New()
            List = New ObservableCollection(Of TestData)()
            PopulateData()
        End Sub

        Private Sub PopulateData()
            For i As Integer = 0 To 30 - 1
                Dim t = New TestData(i, "Element" & i.ToString())
                List.Add(t)
            Next
        End Sub
    End Class

    Public Class TestData
        Implements INotifyPropertyChanged, IDataErrorInfo

        Private numberField As Integer

        Private textField As String

        Public Sub New(ByVal number As Integer, ByVal text As String)
            numberField = number
            textField = text
        End Sub

        Public Property Number As Integer
            Get
                Return numberField
            End Get

            Set(ByVal value As Integer)
                If numberField = value Then Return
                numberField = value
                NotifyChanged("Number")
            End Set
        End Property

        Public Property Text As String
            Get
                Return textField
            End Get

            Set(ByVal value As String)
                If Equals(textField, value) Then Return
                textField = value
                NotifyChanged("Text")
            End Set
        End Property

        Public ReadOnly Property [Error] As String Implements IDataErrorInfo.[Error]
            Get
                Dim result As String = String.Empty
                If Number >= 10 AndAlso Number < 20 Then result = "Incorrect value!"
                Return result
            End Get
        End Property

        Default Public ReadOnly Property Item(ByVal columnName As String) As String Implements IDataErrorInfo.Item
            Get
                Dim result As String = String.Empty
                Select Case columnName
                    Case "Number"
                        If Number > 20 AndAlso Number <= 40 Then result = String.Format("The current number is less than {0} !", Number.ToString())
                End Select

                Return result
            End Get
        End Property

        Private Sub NotifyChanged(ByVal propertyName As String)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub

#Region "INotifyPropertyChanged Members"
        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged
#End Region
    End Class
End Namespace

File: CS/WpfApplication25/ViewModel.cs
using DevExpress.XtraEditors.DXErrorProvider;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WpfApplication25
{
    public class ViewModel
    {
        public ObservableCollection<TestData> List
        {
            get;
            set;
        }

        public ViewModel()
        {
            List = new ObservableCollection<TestData>();
            PopulateData();
        }

        private void PopulateData()
        {
            for (int i = 0; i < 30; i++)
            {
                var t = new TestData(i, "Element" + i.ToString());
                List.Add(t);
            }
        }
    }

    public class TestData : INotifyPropertyChanged, IDataErrorInfo
    {
        int number;
        string text;

        public TestData(int number, string text)
        {
            this.number = number;
            this.text = text;
        }

        public int Number
        {
            get { return number; }
            set
            {
                if (number == value)
                    return;
                number = value;
                NotifyChanged("Number");
            }
        }
        public string Text
        {
            get { return text; }
            set
            {
                if (text == value)
                    return;
                text = value;
                NotifyChanged("Text");
            }
        }

        public string Error
        {
            get
            {
                string result = String.Empty;
                if (Number >= 10 && Number < 20)
                    result = "Incorrect value!";
                return result;
            }
        }

        public string this[string columnName]
        {
            get
            {
                string result = String.Empty;
                switch (columnName)
                {
                    case "Number":
                        if (Number > 20 && Number <= 40)
                            result = String.Format("The current number is less than {0} !", Number.ToString());
                        break;
                }
                return result;
            }
        }

        void NotifyChanged(string propertyName)
        {
            if (PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
        }

        #region INotifyPropertyChanged Members
        public event PropertyChangedEventHandler PropertyChanged;
        #endregion
    }
}
